File: jasusi_cli/core/router.py
"""
ScoredRouter v3 — 6-dimensional intent classifier for JasusiCLI v3.2.
No LLM calls. Pure heuristic scoring. O(n) on query length.
"""

import re
from dataclasses import dataclass, field

ROLES = ["developer", "executor", "architect", "researcher", "reviewer"]

# ── Dimension 1: Lexical keyword bags ────────────────────────────────────────
KEYWORD_SCORES: dict[str, list[tuple[str, float]]] = {
    "developer":  [("implement",0.4),("fix",0.3),("refactor",0.4),
                   ("write",0.3),("build",0.3),("debug",0.4),("patch",0.3),
                   ("function",0.2),("class",0.2),("error",0.2)],
    "executor":   [("bash",0.4),("run",0.3),("deploy",0.4),("execute",0.4),
                   ("install",0.3),("start",0.3),("stop",0.3),("curl",0.4),
                   ("git",0.3),("orchestrate",0.4),("pipeline",0.2)],
    "architect":  [("architecture",0.4),("design",0.3),("pattern",0.3),
                   ("trade-off",0.4),("tradeoff",0.4),("structure",0.3),
                   ("scaffold",0.3),("schema",0.3),("system",0.2)],
    "researcher": [("docs",0.4),("explain",0.3),("guide",0.3),
                   ("documentation",0.4),("research",0.4),("reference",0.3)],
    "reviewer":   [("review",0.4),("audit",0.4),("security",0.3),
                   ("lint",0.4),("scan",0.3),("vulnerability",0.4),
                   ("check",0.2),("validate",0.3)],
}

# ── Dimension 2: Question-start tokens → Researcher ─────────────────────────
QUESTION_STARTS = {"what", "why", "how", "where", "explain", "docs"}

# ── Dimension 5: Imperative verbs → Executor (highest weight: +0.6) ─────────
IMPERATIVE_VERBS = {
    "run","deploy","bash","execute","install","start",
    "stop","curl","git","orchestrate"
}

# ── Dimension 6: Semantic complexity conjunctions ────────────────────────────
COMPLEXITY_PATTERNS = [
    r",\s*then\b", r"\band also\b", r"\bbut first\b", r"\bafter that\b",
    r"\bif .{1,30} fails\b", r"\bunless\b", r"\bwhen complete\b",
]


# ── Calibration constants ────────────────────────────────────────────────────

#: Safety-ordered preference used to break near-ties deterministically.
TIE_BREAK_ORDER = ["executor", "developer", "architect", "researcher", "reviewer"]

#: Score gap below which the top two routes are treated as indistinguishable.
AMBIGUITY_MARGIN = 0.05

#: Minimum evidence required to accept a classification at all.
CONFIDENCE_THRESHOLD = 0.45

#: Deterministic fallback when evidence is insufficient.
DEFAULT_ROLE = "developer"

#: Token pressure at which routing yields to compaction.
COMPACTION_TOKEN_THRESHOLD = 50_000


@dataclass
class RouteScore:
    role: str
    score: float
    dimensions: dict[str, object] = field(default_factory=dict)


@dataclass(frozen=True)
class RouteEvaluation:
    """A route decision together with the signals that justify it.

    ``margin`` and ``ambiguous`` expose calibration rather than a raw score
    ratio, so callers can distinguish "confidently developer" from "developer
    because nothing else scored".
    """

    role: str
    score: float
    runner_up: str
    runner_up_score: float
    margin: float
    ambiguous: bool
    below_evidence_floor: bool
    dimensions: dict[str, dict[str, object]] = field(default_factory=dict)

    @property
    def confidence(self) -> float:
        """Calibrated confidence in [0, 1].

        Derived from the margin over the runner-up, not from the winner's share
        of the total score: a large absolute score means little when a second
        route scored almost as high.
        """
        if self.below_evidence_floor:
            return MIN_CONFIDENCE
        return max(MIN_CONFIDENCE, min(1.0, MIN_CONFIDENCE + self.margin))


#: Floor reported for fallback decisions, so callers never see a bare 0.0.
MIN_CONFIDENCE = 0.4


def evaluate_query(query: str, token_count: int = 0) -> RouteEvaluation:
    """
    Score a query across six heuristic dimensions.
    Compaction check is handled by :func:`evaluate` before calling this.
    """
    q = query.strip()
    q_lower = q.lower()
    first_token = q_lower.split()[0] if q_lower.split() else ""
    scores: dict[str, float] = {r: 0.0 for r in ROLES}
    dims: dict[str, dict[str, object]] = {r: {} for r in ROLES}

    # ── Dim 1: Lexical match (max +0.5 per role) ─────────────────────────────
    for role, pairs in KEYWORD_SCORES.items():
        total = 0.0
        for kw, weight in pairs:
            if kw in q_lower:
                total = min(total + weight, 0.5)
        scores[role] += total
        dims[role]["lexical"] = round(total, 3)

    # ── Dim 2: Question heuristic → Researcher (+0.5) ────────────────────────
    if first_token in QUESTION_STARTS:
        scores["researcher"] += 0.5
        dims["researcher"]["question_heuristic"] = 0.5

    # ── Dim 3: Markdown artifact → Developer (+0.4) ──────────────────────────
    if "```" in q:
        scores["developer"] += 0.4
        dims["developer"]["markdown_artifact"] = 0.4

    # ── Dim 4: Length velocity → Architect ───────────────────────────────────
    char_count = len(q)
    if char_count > 400:
        scores["architect"] += 0.4
        scores["developer"] = max(0.0, scores["developer"] - 0.1)
        dims["architect"]["length_velocity"] = 0.4
    elif char_count > 200:
        scores["architect"] += 0.3
        dims["architect"]["length_velocity"] = 0.3

    # ── Dim 5: Imperative verb → Executor (+0.6, highest weight) ─────────────
    if first_token in IMPERATIVE_VERBS:
        scores["executor"] += 0.6
        dims["executor"]["imperative_verb"] = 0.6

    # ── Dim 6: Semantic complexity → Architect / Developer split ─────────────
    clause_count = sum(
        1 for p in COMPLEXITY_PATTERNS if re.search(p, q_lower)
    )
    if clause_count >= 4:
        scores["architect"] += 0.4
        scores["developer"] = max(0.0, scores["developer"] - 0.2)
        dims["architect"]["semantic_complexity"] = 0.4
    elif clause_count >= 2:
        scores["architect"] += 0.2
        dims["architect"]["semantic_complexity"] = 0.2

    # ── Tie-break: within 0.05 → prefer by safety hierarchy ─────────────────
    ranked = sorted(
        scores.items(),
        key=lambda x: (-x[1], TIE_BREAK_ORDER.index(x[0])),
    )
    best_role, best_score = ranked[0]
    second_role, second_score = ranked[1]

    ambiguous = (best_score - second_score) <= AMBIGUITY_MARGIN
    if ambiguous:
        for preferred in TIE_BREAK_ORDER:
            if preferred in (best_role, second_role):
                best_role = preferred
                break

    below_evidence_floor = best_score < CONFIDENCE_THRESHOLD
    if below_evidence_floor:
        best_role = DEFAULT_ROLE

    return RouteEvaluation(
        role=best_role,
        score=best_score,
        runner_up=second_role,
        runner_up_score=second_score,
        margin=best_score - second_score,
        ambiguous=ambiguous,
        below_evidence_floor=below_evidence_floor,
        dimensions=dims,
    )


def score_query(query: str, token_count: int = 0) -> str:
    """Return the winning role name for a query."""
    return evaluate_query(query, token_count).role


def route(query: str, token_count: int = 0) -> str:
    """Public entry point. Returns role name string."""
    return evaluate(query, token_count).role


def evaluate(query: str, token_count: int = 0) -> RouteEvaluation:
    """Full route decision including calibration signals.

    This is the single scoring implementation for the Python side. Both
    ``core.router`` callers and ``routing.scored_router`` resolve through it, so
    the CLI and the web adapter cannot reach different conclusions for the same
    prompt.
    """
    if token_count >= COMPACTION_TOKEN_THRESHOLD:
        return RouteEvaluation(
            role="compaction",
            score=1.0,
            runner_up=DEFAULT_ROLE,
            runner_up_score=0.0,
            margin=1.0,
            ambiguous=False,
            below_evidence_floor=False,
            dimensions={},
        )
    return evaluate_query(query, token_count)

File: rust/crates/runtime/src/router.rs
use regex::Regex;
use std::collections::BTreeMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum AgentRole {
    Developer,
    Executor,
    Architect,
    Researcher,
    Reviewer,
    Compaction,
}

impl AgentRole {
    #[must_use]
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Developer => "developer",
            Self::Executor => "executor",
            Self::Architect => "architect",
            Self::Researcher => "researcher",
            Self::Reviewer => "reviewer",
            Self::Compaction => "compaction",
        }
    }

    #[must_use]
    pub fn parse(s: &str) -> Option<Self> {
        match s.trim().to_ascii_lowercase().as_str() {
            "developer" | "dev" => Some(Self::Developer),
            "executor" | "exec" => Some(Self::Executor),
            "architect" | "arch" => Some(Self::Architect),
            "researcher" | "research" => Some(Self::Researcher),
            "reviewer" | "review" => Some(Self::Reviewer),
            "compaction" | "compact" => Some(Self::Compaction),
            _ => None,
        }
    }
}

pub const ROLES: [AgentRole; 5] = [
    AgentRole::Developer,
    AgentRole::Executor,
    AgentRole::Architect,
    AgentRole::Researcher,
    AgentRole::Reviewer,
];

pub const TIE_BREAK_ORDER: [AgentRole; 5] = [
    AgentRole::Executor,
    AgentRole::Developer,
    AgentRole::Architect,
    AgentRole::Researcher,
    AgentRole::Reviewer,
];

fn keyword_scores() -> &'static [(AgentRole, &'static [(&'static str, f64)])] {
    &[
        (
            AgentRole::Developer,
            &[
                ("implement", 0.4),
                ("fix", 0.3),
                ("refactor", 0.4),
                ("write", 0.3),
                ("build", 0.3),
                ("debug", 0.4),
                ("patch", 0.3),
                ("function", 0.2),
                ("class", 0.2),
                ("error", 0.2),
            ],
        ),
        (
            AgentRole::Executor,
            &[
                ("bash", 0.4),
                ("run", 0.3),
                ("deploy", 0.4),
                ("execute", 0.4),
                ("install", 0.3),
                ("start", 0.3),
                ("stop", 0.3),
                ("curl", 0.4),
                ("git", 0.3),
                ("orchestrate", 0.4),
                ("pipeline", 0.2),
            ],
        ),
        (
            AgentRole::Architect,
            &[
                ("architecture", 0.4),
                ("design", 0.3),
                ("pattern", 0.3),
                ("trade-off", 0.4),
                ("tradeoff", 0.4),
                ("structure", 0.3),
                ("scaffold", 0.3),
                ("schema", 0.3),
                ("system", 0.2),
            ],
        ),
        (
            AgentRole::Researcher,
            &[
                ("docs", 0.4),
                ("explain", 0.3),
                ("guide", 0.3),
                ("documentation", 0.4),
                ("research", 0.4),
                ("reference", 0.3),
            ],
        ),
        (
            AgentRole::Reviewer,
            &[
                ("review", 0.4),
                ("audit", 0.4),
                ("security", 0.3),
                ("lint", 0.4),
                ("scan", 0.3),
                ("vulnerability", 0.4),
                ("check", 0.2),
                ("validate", 0.3),
            ],
        ),
    ]
}

const QUESTION_STARTS: [&str; 6] = ["what", "why", "how", "where", "explain", "docs"];

const IMPERATIVE_VERBS: [&str; 10] = [
    "run",
    "deploy",
    "bash",
    "execute",
    "install",
    "start",
    "stop",
    "curl",
    "git",
    "orchestrate",
];

const COMPLEXITY_PATTERNS: [&str; 7] = [
    r",\s*then\b",
    r"\band also\b",
    r"\bbut first\b",
    r"\bafter that\b",
    r"\bif .{1,30} fails\b",
    r"\bunless\b",
    r"\bwhen complete\b",
];

#[derive(Debug, Clone)]
pub struct RouteResult {
    pub winning_role: AgentRole,
    pub score: f64,
    pub scores: BTreeMap<String, f64>,
}

const CONFIDENCE_THRESHOLD: f64 = 0.45;

#[must_use]
#[allow(clippy::too_many_lines)]
pub fn score_query(query: &str) -> RouteResult {
    let q = query.trim();
    let q_lower = q.to_lowercase();
    let first_token = q_lower.split_whitespace().next().unwrap_or("");

    let mut scores: BTreeMap<AgentRole, f64> = ROLES.iter().map(|&r| (r, 0.0)).collect();

    // Dim 1: Lexical match (max +0.5 per role)
    for &(role, pairs) in keyword_scores() {
        let mut total = 0.0;
        for &(kw, weight) in pairs {
            if q_lower.contains(kw) {
                total = (total + weight).min(0.5);
            }
        }
        if let Some(s) = scores.get_mut(&role) {
            *s += total;
        }
    }

    // Dim 2: Question start token -> Researcher (+0.5)
    if QUESTION_STARTS.contains(&first_token) {
        if let Some(s) = scores.get_mut(&AgentRole::Researcher) {
            *s += 0.5;
        }
    }

    // Dim 3: Markdown artifact -> Developer (+0.4)
    if q.contains("```") {
        if let Some(s) = scores.get_mut(&AgentRole::Developer) {
            *s += 0.4;
        }
    }

    // Dim 4: Length velocity -> Architect (+0.3 or +0.4, Dev penalty)
    let char_count = q.chars().count();
    if char_count > 400 {
        if let Some(s) = scores.get_mut(&AgentRole::Architect) {
            *s += 0.4;
        }
        if let Some(s) = scores.get_mut(&AgentRole::Developer) {
            *s = (*s - 0.1).max(0.0);
        }
    } else if char_count > 200 {
        if let Some(s) = scores.get_mut(&AgentRole::Architect) {
            *s += 0.3;
        }
    }

    // Dim 5: Imperative verb -> Executor (+0.6)
    if IMPERATIVE_VERBS.contains(&first_token) {
        if let Some(s) = scores.get_mut(&AgentRole::Executor) {
            *s += 0.6;
        }
    }

    // Dim 6: Semantic complexity -> Architect (+0.2 or +0.4, Dev penalty)
    let mut clause_count = 0;
    for pat in &COMPLEXITY_PATTERNS {
        if let Ok(re) = Regex::new(pat) {
            if re.is_match(&q_lower) {
                clause_count += 1;
            }
        }
    }

    if clause_count >= 4 {
        if let Some(s) = scores.get_mut(&AgentRole::Architect) {
            *s += 0.4;
        }
        if let Some(s) = scores.get_mut(&AgentRole::Developer) {
            *s = (*s - 0.2).max(0.0);
        }
    } else if clause_count >= 2 {
        if let Some(s) = scores.get_mut(&AgentRole::Architect) {
            *s += 0.2;
        }
    }

    // Convert BTreeMap to sorted Vec by score desc, tie-break hierarchy asc
    let mut ranked: Vec<(AgentRole, f64)> = scores.into_iter().collect();
    ranked.sort_by(|a, b| {
        b.1.partial_cmp(&a.1)
            .unwrap_or(std::cmp::Ordering::Equal)
            .then_with(|| {
                let pos_a = TIE_BREAK_ORDER
                    .iter()
                    .position(|&r| r == a.0)
                    .unwrap_or(usize::MAX);
                let pos_b = TIE_BREAK_ORDER
                    .iter()
                    .position(|&r| r == b.0)
                    .unwrap_or(usize::MAX);
                pos_a.cmp(&pos_b)
            })
    });

    let (mut best_role, best_score) = ranked[0];
    let (_, second_score) = ranked[1];

    // Tie-break: within 0.05 -> prefer by safety hierarchy
    if (best_score - second_score) <= 0.05 {
        for &preferred in &TIE_BREAK_ORDER {
            if ranked.iter().take(2).any(|&(r, _)| r == preferred) {
                best_role = preferred;
                break;
            }
        }
    }

    // Confidence floor: below 0.45 -> default Developer
    let winning_role = if best_score < CONFIDENCE_THRESHOLD {
        AgentRole::Developer
    } else {
        best_role
    };

    let scores_map = ranked
        .into_iter()
        .map(|(r, s)| (r.as_str().to_string(), s))
        .collect();

    RouteResult {
        winning_role,
        score: best_score,
        scores: scores_map,
    }
}

#[must_use]
pub fn route(query: &str, token_count: usize) -> AgentRole {
    if token_count >= 50_000 {
        return AgentRole::Compaction;
    }
    score_query(query).winning_role
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn routes_compaction_on_large_token_count() {
        assert_eq!(route("fix bug", 50_000), AgentRole::Compaction);
        assert_eq!(route("fix bug", 60_000), AgentRole::Compaction);
    }

    #[test]
    fn routes_executor_on_imperative_verbs() {
        assert_eq!(route("run cargo test", 100), AgentRole::Executor);
        assert_eq!(
            route("deploy the application to staging", 100),
            AgentRole::Executor
        );
        assert_eq!(route("bash ls -la", 100), AgentRole::Executor);
    }

    #[test]
    fn routes_researcher_on_questions() {
        assert_eq!(
            route("what is the purpose of this file?", 100),
            AgentRole::Researcher
        );
        assert_eq!(
            route("explain how authentication works", 100),
            AgentRole::Researcher
        );
    }

    #[test]
    fn routes_developer_on_code_modifications() {
        assert_eq!(
            route(
                "implement a new feature in rust with function and class",
                100
            ),
            AgentRole::Developer
        );
        assert_eq!(
            route("fix the bug in function process_data", 100),
            AgentRole::Developer
        );
    }

    #[test]
    fn routes_reviewer_on_audit_query() {
        assert_eq!(
            route(
                "review and audit security vulnerability in the codebase",
                100
            ),
            AgentRole::Reviewer
        );
    }

    #[test]
    fn routes_architect_on_large_query_and_complexity() {
        let long_query = "architecture design pattern system structure ".repeat(15);
        assert_eq!(route(&long_query, 100), AgentRole::Architect);
    }

    #[test]
    fn defaults_to_developer_on_low_confidence() {
        assert_eq!(route("xyz", 100), AgentRole::Developer);
    }
}
